File: Information_Aggregation/backend/app/schemas/match.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class MatchRequirements(BaseModel):
    """智能匹配条件"""

    platform: str | None = None
    follower_min: int | None = Field(default=None, ge=0)
    follower_max: int | None = Field(default=None, ge=0)
    required_tag_ids: list[int] | None = Field(default=None, description="必须全部命中的标签")
    preferred_tag_ids: list[int] | None = Field(default=None, description="优先匹配的标签（加分）")
    agency_id: int | None = None
    engagement_rate_min: float | None = Field(default=None, ge=0, le=1)
    keyword: str | None = None
    must_have_contact: bool = False
    limit: int = Field(default=50, ge=1, le=200)


class MatchRequestCreate(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    requirements: MatchRequirements


class MatchReasonDetail(BaseModel):
    dimension: str
    score: float
    max_score: float
    note: str


class MatchReasonOut(BaseModel):
    summary: str
    details: list[MatchReasonDetail] = []


class MatchInfluencerBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    platform: str
    platform_uid: str
    nickname: str | None
    avatar_url: str | None
    follower_count: int
    engagement_rate: float | None
    agency_name: str | None = None
    tags: list[str] = []


class MatchResultOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    request_id: int
    influencer_id: int
    match_score: float | None
    rank_order: int | None
    reason: MatchReasonOut | None
    is_selected: bool
    influencer: MatchInfluencerBrief | None = None


class MatchRequestOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    title: str | None
    requirements: dict
    status: str
    result_count: int | None
    created_at: datetime
    selected_count: int = 0


class MatchRequestDetailOut(MatchRequestOut):
    top_results: list[MatchResultOut] = []


class MatchSelectionUpdate(BaseModel):
    result_ids: list[int] = Field(..., min_length=0, max_length=200)
    selected: bool = True

File: Information_Aggregation/frontend/src/api/match.ts
import axios from 'axios'
import request, { type ApiResponse, type PageResult } from './request'

export interface MatchRequirements {
  platform?: string
  follower_min?: number
  follower_max?: number
  required_tag_ids?: number[]
  preferred_tag_ids?: number[]
  agency_id?: number
  engagement_rate_min?: number
  keyword?: string
  must_have_contact?: boolean
  limit?: number
}

export interface MatchReasonDetail {
  dimension: string
  score: number
  max_score: number
  note: string
}

export interface MatchReason {
  summary: string
  details: MatchReasonDetail[]
}

export interface MatchInfluencerBrief {
  id: number
  platform: string
  platform_uid: string
  nickname: string | null
  avatar_url: string | null
  follower_count: number
  engagement_rate: number | null
  agency_name: string | null
  tags: string[]
}

export interface MatchResult {
  id: number
  request_id: number
  influencer_id: number
  match_score: number | null
  rank_order: number | null
  reason: MatchReason | null
  is_selected: boolean
  influencer: MatchInfluencerBrief | null
}

export interface MatchRequest {
  id: number
  user_id: number
  title: string | null
  requirements: MatchRequirements
  status: string
  result_count: number | null
  created_at: string
  selected_count: number
}

export interface MatchRequestDetail extends MatchRequest {
  top_results: MatchResult[]
}

export function createMatchRequest(data: { title?: string; requirements: MatchRequirements }) {
  return request.post<any, ApiResponse<MatchRequest>>('/match/requests', data)
}

export function getMatchRequests(params: { page?: number; page_size?: number }) {
  return request.get<any, ApiResponse<PageResult<MatchRequest>>>('/match/requests', { params })
}

export function getMatchRequest(id: number) {
  return request.get<any, ApiResponse<MatchRequestDetail>>(`/match/requests/${id}`)
}

export function getMatchResults(
  id: number,
  params: { page?: number; page_size?: number; selected_only?: boolean }
) {
  return request.get<any, ApiResponse<PageResult<MatchResult>>>(`/match/requests/${id}/results`, {
    params,
  })
}

export function updateMatchSelection(
  id: number,
  data: { result_ids: number[]; selected: boolean }
) {
  return request.put<any, ApiResponse<{ updated: number }>>(`/match/requests/${id}/selection`, data)
}

export function deleteMatchRequest(id: number) {
  return request.delete<any, ApiResponse<null>>(`/match/requests/${id}`)
}

export async function downloadMatchExport(requestId: number, selectedOnly = false) {
  const res = await axios.get(`/api/v1/match/requests/${requestId}/export`, {
    params: { selected_only: selectedOnly },
    responseType: 'blob',
    headers: { Authorization: `Bearer ${localStorage.getItem('token') || ''}` },
  })
  const url = window.URL.createObjectURL(res.data)
  const link = document.createElement('a')
  link.href = url
  link.download = `match_${requestId}.xlsx`
  link.click()
  window.URL.revokeObjectURL(url)
}

export const MATCH_STATUS_MAP: Record<string, { label: string; type: string }> = {
  pending: { label: '等待中', type: 'info' },
  running: { label: '匹配中', type: 'warning' },
  completed: { label: '已完成', type: 'success' },
  failed: { label: '失败', type: 'danger' },
}

export const DIMENSION_LABELS: Record<string, string> = {
  tags: '标签',
  followers: '粉丝量',
  engagement: '互动率',
  agency: '机构',
  profile: '档案完整度',
}
